Add: tab audible behaviour

PikaDude opened the pull request:
This PR adds a customisable function that will trigger whenever a tab begins to play audio, such as lowering the extension's music volume or pausing it. Also includes some other small changes and fixes.

Closes animal-crossing-music-extension/Animal-Crossing-Music-Extension#7
![image](https://user-images.githubusercontent.com/18642415/72957827-20859d80-3df9-11ea-95eb-02532919f0ba.png)

PikaDude committed:
Add: tab audible behaviour

JordiGarcL commented:
Will check this now

jordigarcl commented on scripts/background/TabAudioHandler.js:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    function checkTabs() {
+        // A tab can be muted and still be "audible"
```
I'm curious, how is that so?

jordigarcl commented on scripts/background/AudioManager.js:
```diff
@@ -164,6 +169,54 @@ function AudioManager(addEventListener, isTownTune) {
+				reducedVolume = false;
+				audio.volume = setVolume;
+			}
+			window.notify("tabAudio", [true, tabAudio, reduceValue]);
```
Current logic introduces following bug:
If music **from the extension** is playing (and no other tabs are playing music), when a 'Tab Audio Behaviour' option is switched to 'reduce' /'pause', extension music is reduced/paused. 

That is, switching to 'pause' will effectively pause the extension even though there are no other sounds.

JordiGarcL commented:
Wow, great work.

Although I think the nested `if`s on the AudioManager's `tabAudio` event listener could be refactored (maybe leveraging JS's switch statement), code looks good!

I did found a bug that occurs when switching options with the extension enabled and no external music playing. As it is not a critical bug (and is trivial to fix) I'm unblocking the PR!  :+1

JordiGarcL commented:
Wait, just found a more important bug which should be  strongly related to the first one.
For some reason manually pausing extension music and then triggering the `chrome.tabs.onUpdated` callback (for example by reloading or creating/deleting a tab) makes the extension play. This only happens when the 'pause music' option is selected.

jordigarcl reviewed (approved):
I think the 'music re-starting automatically when updating tabs' is critical enough to be addressed before merging

jordigarcl reviewed (changes requested):
Ooops

PikaDude committed:
Fix: various bugs pointed out in #38

PikaDude commented:
The `tabAudio` event listener is purposely left as it is to make it a bit more human-readable and easier to follow. I found that refactoring it made it hard to read and follow. Also nice catch on those bugs! They were all related to a small mistake that I've now fixed. I also noticed and fixed another issue that was causing this function to try and pause the music while it was still loading, throwing errors that didn't really mean anything, but errors nonetheless.

jordigarcl reviewed (approved):
Last change fixed the music stopping bug 🥳

PikaDude committed:
Merge branch 'develop' into feature/react-tab-audio
